rust/azure_iot_operations_services/src/azure_device_registry/client.rs

```diff
@@ -1,2 +1,1529 @@
+                if self
+                    .device_update_notification_dispatcher
+                    .unregister_receiver(&receiver_id)
+                {
+                    log::debug!(
+                        "Device `{device_name:?}` with inbound endpoint `{inbound_endpoint_name:?}` removed from observed list"
+                    );
+                } else {
+                    log::debug!(
+                        "Device `{device_name:?}` with inbound endpoint `{inbound_endpoint_name:?}` not in observed list"
+                    );
+                }
+                Err(Error(ErrorKind::from(e)))
+            }
+        }
+    }
+
+    /// Stops observation of a device's updates from the Azure Device Registry service.
+    ///
+    /// # Arguments
+    /// * `device_name` - The name of the device.
+    /// * `inbound_endpoint_name` - The name of the inbound endpoint.
+    /// * `timeout` - The duration until the client stops waiting for a response to the request, it is rounded up to the nearest second.
+    ///
+    /// Returns `Ok(())` if the device updates are no longer being observed.
+    ///
+    /// # Errors
+    /// [`struct@Error`] of kind [`InvalidRequestArgument`](ErrorKind::InvalidRequestArgument)
+    /// if timeout is 0 or > `u32::max`.
+    ///
+    /// [`struct@Error`] of kind [`AIOProtocolError`](ErrorKind::AIOProtocolError) if:
+    /// - device or inbound endpoint names are invalid.
+    /// - there are any underlying errors from the AIO RPC protocol.
+    ///
+    /// [`struct@Error`] of kind [`ObservationError`](ErrorKind::ObservationError)
+    /// if the observation was not successful.
+    pub async fn unobserve_device_update_notifications(
+        &self,
+        device_name: String,
+        inbound_endpoint_name: String,
+        timeout: Duration,
+    ) -> Result<(), Error> {
+        let unobserve_payload =
+            adr_name_gen::SetNotificationPreferenceForDeviceUpdatesRequestPayload {
+                notification_preference_request: adr_name_gen::NotificationPreference::Off,
+            };
+
+        let unobserve_request =
+            adr_name_gen::SetNotificationPreferenceForDeviceUpdatesRequestBuilder::default()
+                .payload(unobserve_payload)
+                .map_err(ErrorKind::from)?
+                .topic_tokens(Self::get_topic_tokens(
+                    device_name.clone(),
+                    inbound_endpoint_name.clone(),
+                ))
+                .timeout(timeout)
+                .build()
+                .map_err(ErrorKind::from)?;
+        let response = self
+            .notify_on_device_update_command_invoker
+            .invoke(unobserve_request)
+            .await
+            .map_err(ErrorKind::from)?;
+        match response.payload.notification_preference_response {
+            adr_name_gen::NotificationPreferenceResponse::Accepted => {
+                let receiver_id = Self::hash_device_endpoint(&device_name, &inbound_endpoint_name);
```

Comment:
Maybe a note for the future but I think the dispatcher could have a generic as the key name as long as it implements the `Hash` trait. If you do that you won't need to implement the hash and unhash functions, you'd just use a struct in this case like so:

``` Rust
#[derive(Hash)]
struct device_endpoint {
  device_name: String, 
  inbound_endpoint_name: String,
}
```

Reply:
Come to think of it, this also looks like `DeviceRef`, so we could potentially just have `DeviceRef` derive that trait